File: connectors/microsoft/ms_connector/connector.py
"""Main MicrosoftConnector class."""

import logging
from typing import Any
from starlette.responses import Response

from omni_connector import Connector, SearchOperator, SyncContext
from omni_connector.models import (
    ActionDefinition,
    ActionResponse,
    OAuthManifestConfig,
    OAuthScopeSet,
)

from .auth import MSGraphAuth, parse_ms_credentials
from .graph_client import AuthenticationError, GraphAPIError, GraphClient
from .syncers.calendar import CalendarSyncer
from .syncers.mail import MailSyncer
from .syncers.onedrive import OneDriveSyncer
from .syncers.sharepoint import SharePointSyncer
from .syncers.teams import TeamsSyncer

logger = logging.getLogger(__name__)

SOURCE_TYPE_TO_SYNCER = {
    "one_drive": "onedrive",
    "share_point": "sharepoint",
    "outlook": "mail",
    "outlook_calendar": "calendar",
    "ms_teams": "teams",
}


class MicrosoftConnector(Connector):
    """Microsoft 365 connector for Omni.

    Syncs OneDrive files, Outlook mail, Outlook calendar events,
    and SharePoint document libraries via the Microsoft Graph API.
    Each source type maps to exactly one syncer.
    """

    @property
    def name(self) -> str:
        return "microsoft"

    @property
    def display_name(self) -> str:
        return "Microsoft"

    @property
    def version(self) -> str:
        return "1.0.0"

    @property
    def source_types(self) -> list[str]:
        return ["one_drive", "share_point", "outlook", "outlook_calendar", "ms_teams"]

    @property
    def description(self) -> str:
        return "Connect to OneDrive, SharePoint, Outlook mail, calendar, and Teams"

    @property
    def sync_modes(self) -> list[str]:
        return ["full", "incremental"]

    @property
    def search_operators(self) -> list[SearchOperator]:
        return [
            SearchOperator(
                operator="from", attribute_key="sender", value_type="person"
            ),
            SearchOperator(operator="to", attribute_key="to", value_type="person"),
            SearchOperator(operator="cc", attribute_key="cc", value_type="person"),
        ]

    def oauth_config(self) -> OAuthManifestConfig | None:
        # Endpoints below default to the `organizations` tenant; the web layer
        # overrides them with tenant-specific URLs from connector_configs at
        # flow time (admin's tenant_id is materialized into the stored
        # auth/token endpoints during setup).
        return OAuthManifestConfig(
            provider="microsoft",
            auth_endpoint="https://login.microsoftonline.com/organizations/oauth2/v2.0/authorize",
            token_endpoint="https://login.microsoftonline.com/organizations/oauth2/v2.0/token",
            userinfo_endpoint="https://graph.microsoft.com/v1.0/me",
            userinfo_email_field="userPrincipalName",
            identity_scopes=[
                "openid",
                "profile",
                "email",
                "offline_access",
                "User.Read",
            ],
            scopes={
                "one_drive": OAuthScopeSet(
                    read=["Files.Read.All"],
                    write=["Files.ReadWrite.All"],
                ),
                "share_point": OAuthScopeSet(
                    read=["Sites.Read.All", "Files.Read.All"],
                    write=["Sites.ReadWrite.All", "Files.ReadWrite.All"],
                ),
                "outlook": OAuthScopeSet(
                    read=["Mail.Read"],
                    write=["Mail.ReadWrite", "Mail.Send"],
                ),
                "outlook_calendar": OAuthScopeSet(
                    read=["Calendars.Read"],
                    write=["Calendars.ReadWrite"],
                ),
                "ms_teams": OAuthScopeSet(
                    read=[
                        "Channel.ReadBasic.All",
                        "ChannelMessage.Read.All",
                        "Team.ReadBasic.All",
                    ],
                    write=["ChannelMessage.Send"],
                ),
            },
            scope_separator=" ",
            extra_auth_params={},
        )

    @property
    def actions(self) -> list[ActionDefinition]:
        return [
            ActionDefinition(
                name="search_users",
                description="Search Microsoft 365 users by name or email",
                input_schema={
                    "type": "object",
                    "properties": {
                        "query": {"type": "string", "description": "Search query"},
                    },
                    "required": ["query"],
                },
                admin_only=True,
            ),
            ActionDefinition(
                name="fetch_file",
                description="Download a file from OneDrive or SharePoint",
                mode="read",
                input_schema={
                    "type": "object",
                    "properties": {
                        "file_id": {
                            "type": "string",
                            "description": "External file ID (e.g. onedrive:driveId:itemId)",
                        },
                    },
                    "required": ["file_id"],
                },
            ),
        ]

    async def execute_action(
        self,
        action: str,
        params: dict[str, Any],
        credentials: dict[str, Any],
    ) -> Response:
        if action == "search_users":
            return await self._action_search_users(params, credentials)
        elif action == "fetch_file":
            return await self._action_fetch_file(params, credentials)
        return ActionResponse.not_supported(action).to_response(status_code=404)

    async def _action_search_users(
        self,
        params: dict[str, Any],
        credentials: dict[str, Any],
    ) -> Response:
        query = params.get("query", "").strip()
        if not query:
            return ActionResponse.success({"users": []}).to_response()

        try:
            raw_creds = credentials.get("credentials", credentials)
            auth = MSGraphAuth.from_credentials(parse_ms_credentials(raw_creds))
            client = GraphClient(auth)
            try:
                users = await client.search_users(query, limit=20)
                return ActionResponse.success({"users": users}).to_response()
            finally:
                await client.close()
        except Exception as e:
            logger.exception("search_users action failed")
            return ActionResponse.failure(str(e)).to_response(status_code=500)

    async def _action_fetch_file(
        self,
        params: dict[str, Any],
        credentials: dict[str, Any],
    ) -> Response:
        file_id = params.get("file_id", "").strip()
        if not file_id:
            return ActionResponse.failure(
                "Missing required parameter: file_id"
            ).to_response(status_code=400)

        # Parse external_id: onedrive:{drive_id}:{item_id} or sharepoint:{site_id}:{drive_id}:{item_id}
        parts = file_id.split(":")
        if parts[0] == "onedrive" and len(parts) == 3:
            drive_id = parts[1]
            item_id = parts[2]
        elif parts[0] == "sharepoint" and len(parts) == 4:
            drive_id = parts[2]
            item_id = parts[3]
        else:
            return ActionResponse.failure(
                f"Invalid file_id format: {file_id}. "
                "Expected onedrive:{{drive_id}}:{{item_id}} or "
                "sharepoint:{{site_id}}:{{drive_id}}:{{item_id}}"
            ).to_response(status_code=400)

        try:
            raw_creds = credentials.get("credentials", credentials)
            auth = MSGraphAuth.from_credentials(parse_ms_credentials(raw_creds))
            client = GraphClient(auth)
            try:
                metadata = await client.get_drive_item_metadata(drive_id, item_id)
                file_name = metadata.get("name", "download")
                mime_type = metadata.get("file", {}).get(
                    "mimeType", "application/octet-stream"
                )

                data = await client.get_binary(
                    f"/drives/{drive_id}/items/{item_id}/content"
                )

                logger.info(
                    "fetch_file: returning '%s' (%d bytes, %s)",
                    file_name,
                    len(data),
                    mime_type,
                )

                return Response(
                    content=data,
                    media_type=mime_type,
                    headers={
                        "X-File-Name": file_name,
                        "Content-Length": str(len(data)),
                    },
                )
            finally:
                await client.close()
        except Exception as e:
            logger.exception("fetch_file action failed")
            return ActionResponse.failure(str(e)).to_response(status_code=500)

    async def sync(
        self,
        source_config: dict[str, Any],
        credentials: dict[str, Any],
        checkpoint: dict[str, Any] | None,
        ctx: SyncContext,
    ) -> None:
        try:
            auth = MSGraphAuth.from_credentials(parse_ms_credentials(credentials))
        except ValueError as e:
            await ctx.fail(str(e))
            return

        graph_base_url = source_config.get("graph_base_url")
        client = (
            GraphClient(auth, base_url=graph_base_url)
            if graph_base_url
            else GraphClient(auth)
        )

        try:
            await client.test_connection()
        except AuthenticationError as e:
            await ctx.fail(f"Authentication failed: {e}")
            return
        except GraphAPIError as e:
            await ctx.fail(f"Connection test failed: {e}")
            return

        syncer_key = SOURCE_TYPE_TO_SYNCER.get(ctx.source_type or "")
        if syncer_key is None:
            await ctx.fail(f"Unknown source type: {ctx.source_type}")
            return

        # Sync group memberships and build ID→email caches for permission resolution
        fetched_groups = await self._sync_groups(client, ctx)
        group_cache = {
            g["id"]: (g.get("mail") or "").lower()
            for g in fetched_groups
            if g.get("mail")
        }

        try:
            all_users = await client.list_users()
        except Exception as e:
            logger.warning("Failed to list users for cache: %s", e)
            all_users = []
        user_cache = {
            u["id"]: (u.get("mail") or u.get("userPrincipalName") or "").lower()
            for u in all_users
            if u.get("mail") or u.get("userPrincipalName")
        }

        syncer = self._create_syncer(syncer_key, source_config)
        checkpoint = checkpoint or {}

        logger.info("Starting Microsoft sync (syncer=%s)", syncer_key)

        try:
            result_checkpoint = await syncer.sync(
                client,
                ctx,
                checkpoint,
                source_config=source_config,
                user_cache=user_cache,
                group_cache=group_cache,
            )
            await ctx.complete(checkpoint=result_checkpoint)
            logger.info(
                "Sync completed: %d scanned, %d emitted",
                ctx.documents_scanned,
                ctx.documents_emitted,
            )
        except AuthenticationError as e:
            logger.error("Authentication error during sync: %s", e)
            await ctx.fail(f"Authentication failed: {e}")
        except Exception as e:
            logger.exception("Sync failed with unexpected error")
            await ctx.fail(str(e))
        finally:
            await client.close()

    async def _sync_groups(
        self, client: GraphClient, ctx: SyncContext
    ) -> list[dict[str, Any]]:
        """Sync Entra ID group memberships. Returns fetched groups for cache building."""
        try:
            groups = await client.list_groups()
        except Exception as e:
            logger.warning("Failed to list groups: %s. Skipping group sync.", e)
            return []

        logger.info("Found %d groups, syncing memberships", len(groups))
        total_members = 0

        for group in groups:
            group_id = group["id"]
            group_email = (group.get("mail") or "").lower()
            group_name = group.get("displayName")

            if not group_email:
                continue

            try:
                members = await client.list_group_members(group_id)
            except Exception as e:
                logger.warning(
                    "Failed to list members for group %s: %s", group_email, e
                )
                continue

            member_emails = [
                (m.get("mail") or m.get("userPrincipalName") or "").lower()
                for m in members
            ]
            member_emails = [e for e in member_emails if e]

            total_members += len(member_emails)

            try:
                await ctx.emit_group_membership(
                    group_email=group_email,
                    member_emails=member_emails,
                    group_name=group_name,
                )
            except Exception as e:
                logger.warning(
                    "Failed to emit group membership event for %s: %s",
                    group_email,
                    e,
                )

        logger.info(
            "Group sync complete: %d groups, %d total memberships",
            len(groups),
            total_members,
        )
        return groups

    def _create_syncer(self, syncer_key: str, source_config: dict[str, Any]) -> Any:
        if syncer_key == "onedrive":
            return OneDriveSyncer()
        elif syncer_key == "mail":
            return MailSyncer()
        elif syncer_key == "calendar":
            return CalendarSyncer()
        elif syncer_key == "sharepoint":
            return SharePointSyncer()
        elif syncer_key == "teams":
            return TeamsSyncer()
        raise ValueError(f"Unknown syncer key: {syncer_key}")

File: connectors/atlassian/src/connector.rs
use std::sync::Arc;

use anyhow::Result;
use async_trait::async_trait;
use axum::http::StatusCode;
use axum::response::Response;
use omni_connector_sdk::{
    ActionDefinition, ActionMode, ActionResponse, Connector, SearchOperator, ServiceCredential,
    Source, SourceType, SyncContext, SyncType,
};
use serde_json::{Value as JsonValue, json};
use tracing::info;

use crate::auth::{AtlassianCredentials, AuthManager};
use crate::client::{AtlassianApi, AtlassianClient};
use crate::models::AtlassianSyncCheckpoint;
use crate::sync::SyncManager;

pub struct AtlassianConnector {
    pub sync_manager: Arc<SyncManager>,
}

impl AtlassianConnector {
    pub fn new(sync_manager: Arc<SyncManager>) -> Self {
        Self { sync_manager }
    }
}

#[async_trait]
impl Connector for AtlassianConnector {
    type Config = JsonValue;
    type Credentials = JsonValue;
    type State = AtlassianSyncCheckpoint;

    fn name(&self) -> &'static str {
        "atlassian"
    }

    fn version(&self) -> &'static str {
        "1.0.0"
    }

    fn display_name(&self) -> String {
        "Atlassian".to_string()
    }

    fn description(&self) -> Option<String> {
        Some("Connect to Confluence and Jira using an API token".to_string())
    }

    fn source_types(&self) -> Vec<SourceType> {
        vec![SourceType::Confluence, SourceType::Jira]
    }

    fn sync_modes(&self) -> Vec<SyncType> {
        vec![SyncType::Full, SyncType::Incremental]
    }

    fn actions(&self) -> Vec<ActionDefinition> {
        vec![ActionDefinition {
            name: "search_spaces".to_string(),
            description: "Search Confluence spaces or Jira projects".to_string(),
            input_schema: json!({
                "type": "object",
                "properties": {
                    "query": {
                        "type": "string",
                        "description": "Search query to filter by name or key"
                    },
                    "type": {
                        "type": "string",
                        "description": "Whether to search Confluence spaces or Jira projects"
                    }
                },
                "required": ["type"]
            }),
            mode: ActionMode::Read,
            source_types: Vec::new(),
            admin_only: true,
        }]
    }

    fn search_operators(&self) -> Vec<SearchOperator> {
        vec![
            SearchOperator {
                operator: "status".to_string(),
                attribute_key: "status".to_string(),
                value_type: "text".to_string(),
            },
            SearchOperator {
                operator: "label".to_string(),
                attribute_key: "labels".to_string(),
                value_type: "text".to_string(),
            },
            SearchOperator {
                operator: "project".to_string(),
                attribute_key: "project_key".to_string(),
                value_type: "text".to_string(),
            },
            SearchOperator {
                operator: "assignee".to_string(),
                attribute_key: "assignee".to_string(),
                value_type: "person".to_string(),
            },
        ]
    }

    async fn sync(
        &self,
        source: Source,
        credentials: Option<ServiceCredential>,
        state: Option<Self::State>,
        ctx: SyncContext,
    ) -> Result<()> {
        self.sync_manager
            .run_sync(source, credentials, state, ctx)
            .await
    }

    async fn execute_action(
        &self,
        action: &str,
        params: JsonValue,
        credentials: Option<ServiceCredential>,
    ) -> Result<Response> {
        info!("Action requested: {}", action);
        match action {
            "search_spaces" => Ok(handle_search_spaces(params, credentials)
                .await
                .into_response()),
            _ => Ok(ActionResponse::not_supported(action)
                .into_response_with_status(StatusCode::NOT_FOUND)),
        }
    }

    async fn cancel(&self, _sync_run_id: &str) -> bool {
        // SDK's own cancellation flag (via SyncContext) is the source of truth.
        true
    }
}

pub async fn handle_search_spaces(
    params: JsonValue,
    credentials: Option<ServiceCredential>,
) -> ActionResponse {
    let query = params
        .get("query")
        .and_then(|v| v.as_str())
        .unwrap_or("")
        .to_lowercase();
    let search_type = match params.get("type").and_then(|v| v.as_str()) {
        Some(t) => t.to_string(),
        None => return ActionResponse::failure("Missing required parameter: type"),
    };

    let creds = match credentials {
        Some(c) => c,
        None => return ActionResponse::failure("Atlassian action requires credentials"),
    };

    let domain = match creds.config.get("domain").and_then(|v| v.as_str()) {
        Some(u) => u.to_string(),
        None => return ActionResponse::failure("Missing domain in credentials config"),
    };
    let sa_token = match creds.credentials.get("sa_token").and_then(|v| v.as_str()) {
        Some(t) => t.to_string(),
        None => return ActionResponse::failure("Missing sa_token in credentials"),
    };

    let cloud_id = match AuthManager::new().fetch_cloud_id(&domain).await {
        Ok(id) => id,
        Err(e) => return ActionResponse::failure(format!("Failed to resolve cloud_id: {}", e)),
    };

    let creds = AtlassianCredentials::new(domain, cloud_id, sa_token);
    let client = AtlassianClient::new();

    match search_type.as_str() {
        "confluence" => match client.get_confluence_spaces(&creds).await {
            Ok(spaces) => {
                let results: Vec<JsonValue> = spaces
                    .into_iter()
                    .filter(|s| {
                        s.r#type != "personal"
                            && (query.is_empty()
                                || s.key.to_lowercase().contains(&query)
                                || s.name.to_lowercase().contains(&query))
                    })
                    .map(|s| {
                        json!({
                            "key": s.key,
                            "name": s.name,
                            "type": "confluence"
                        })
                    })
                    .collect();
                ActionResponse::success(json!(results))
            }
            Err(e) => ActionResponse::failure(format!("Failed to fetch Confluence spaces: {}", e)),
        },
        "jira" => match client.get_jira_projects(&creds, &[]).await {
            Ok(projects) => {
                let results: Vec<JsonValue> = projects
                    .into_iter()
                    .filter(|p| {
                        if query.is_empty() {
                            return true;
                        }
                        let key = p
                            .get("key")
                            .and_then(|v| v.as_str())
                            .unwrap_or("")
                            .to_lowercase();
                        let name = p
                            .get("name")
                            .and_then(|v| v.as_str())
                            .unwrap_or("")
                            .to_lowercase();
                        key.contains(&query) || name.contains(&query)
                    })
                    .map(|p| {
                        json!({
                            "key": p.get("key").and_then(|v| v.as_str()).unwrap_or(""),
                            "name": p.get("name").and_then(|v| v.as_str()).unwrap_or(""),
                            "type": "jira"
                        })
                    })
                    .collect();
                ActionResponse::success(json!(results))
            }
            Err(e) => ActionResponse::failure(format!("Failed to fetch Jira projects: {}", e)),
        },
        _ => ActionResponse::failure(format!(
            "Invalid type: {}. Must be 'confluence' or 'jira'",
            search_type
        )),
    }
}
